ygnmi/gnmi.go

```diff
@@ -330,25 +333,49 @@ const (
 )
 
 // populateSetRequest fills a SetResponse for a val and operation type.
-func populateSetRequest(req *gpb.SetRequest, path *gpb.Path, val interface{}, op setOperation, preferShadowPath bool) error {
+func populateSetRequest(req *gpb.SetRequest, path *gpb.Path, val interface{}, op setOperation, preferShadowPath bool, opts ...Option) error {
 	if req == nil {
 		return fmt.Errorf("cannot populate a nil SetRequest")
 	}
+	opt := resolveOpts(opts)
 
 	switch op {
 	case deletePath:
 		req.Delete = append(req.Delete, path)
 	case replacePath, updatePath:
-		// Since the GoStructs are generated using preferOperationalState, we
-		// need to turn on preferShadowPath to prefer marshalling config paths.
-		js, err := ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
-		if err != nil {
-			return fmt.Errorf("could not encode value into JSON format: %w", err)
-		}
 		update := &gpb.Update{
 			Path: path,
-			Val:  &gpb.TypedValue{Value: &gpb.TypedValue_JsonIetfVal{JsonIetfVal: js}},
 		}
+		var typedVal *gpb.TypedValue
+		var err error
+		if opt.preferProto {
+			typedVal, err = ygot.EncodeTypedValue(val, gpb.Encoding_JSON_IETF, &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+			update.Val = typedVal
+		} else {
+			typedVal = &gpb.TypedValue{Value: &gpb.TypedValue_JsonIetfVal{}}
+			// Since the GoStructs are generated using preferOperationalState, we
+			// need to turn on preferShadowPath to prefer marshalling config paths.
+			typedVal.Value.(*gpb.TypedValue_JsonIetfVal).JsonIetfVal, err = ygot.Marshal7951(val, ygot.JSONIndent("  "), &ygot.RFC7951JSONConfig{AppendModuleName: true, PreferShadowPath: preferShadowPath})
+		}
+		update.Val = typedVal
+		if err != nil && opt.setFallback {
+			if m, ok := val.(proto.Message); ok {
+				any, err := anypb.New(m)
+				if err != nil {
+					return fmt.Errorf("failed to marshal proto: %v", err)
+				}
+				update.Val = &gpb.TypedValue{Value: &gpb.TypedValue_AnyVal{AnyVal: any}}
+			} else {
+				b, err := json.Marshal(val)
+				if err != nil {
+					return fmt.Errorf("failed to marshal json: %v", err)
+				}
+				update.Val = &gpb.TypedValue{Value: &gpb.TypedValue_JsonVal{JsonVal: b}}
```

Comment:
I'm not quite sure about doing this -- if the input path's origin is "openconfig" or "" (which is implicitly openconfig), JsonVal should never be used. I'd suggest the entire fallback routine to only apply when the origin is not openconfig.

Reply:
added check for origin